File: pharmion-frontend/pharmion_desktop/lib/services/patient_service.dart
import 'api_service.dart';

class PatientModel {
  final int id;
  final String firstName, lastName, username, email;
  final String phoneNumber, address, cityName;
  final int? cityId;
  final DateTime? dateOfBirth;
  final bool isInsured, isActive;
  final List<String> chronicDiseases;

  const PatientModel({
    required this.id,
    required this.firstName,
    required this.lastName,
    required this.username,
    required this.email,
    required this.phoneNumber,
    required this.address,
    required this.cityName,
    this.cityId,
    this.dateOfBirth,
    required this.isInsured,
    required this.isActive,
    required this.chronicDiseases,
  });

  String get fullName => '$firstName $lastName';

  int get age {
    if (dateOfBirth == null) return 0;
    final now = DateTime.now();
    int age = now.year - dateOfBirth!.year;
    if (now.month < dateOfBirth!.month ||
        (now.month == dateOfBirth!.month && now.day < dateOfBirth!.day)) {
      age--;
    }
    return age;
  }

  factory PatientModel.fromJson(Map<String, dynamic> json) {
    return PatientModel(
      id: json['id'] as int? ?? 0,
      firstName: json['firstName'] as String? ?? '',
      lastName: json['lastName'] as String? ?? '',
      username: json['username'] as String? ?? '',
      email: json['email'] as String? ?? '',
      phoneNumber: json['phoneNumber'] as String? ?? '',
      address: json['address'] as String? ?? '',
      cityName: json['cityName'] as String? ?? '',
      cityId: json['cityId'] as int?,
      dateOfBirth: json['dateOfBirth'] != null
          ? DateTime.tryParse(json['dateOfBirth'])
          : null,
      isInsured: json['isInsured'] as bool? ?? false,
      isActive: json['isActive'] as bool? ?? true,
      chronicDiseases: (json['chronicDiseases'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
    );
  }
}

class PagedPatients {
  final List<PatientModel> items;
  final int totalCount;
  const PagedPatients({required this.items, required this.totalCount});
}

class PatientService {
  static Future<PagedPatients> getPatients({
    int page = 0,
    int pageSize = 10,
    String? name,
    bool? isInsured,
    int? cityId,
  }) async {
    final params = StringBuffer('Patient?includeTotalCount=true&retrieveAll=false');
    params.write('&page=$page&pageSize=$pageSize');
    if (name != null && name.isNotEmpty) params.write('&name=$name');
    if (isInsured != null) params.write('&isInsured=$isInsured');
    if (cityId != null) params.write('&cityId=$cityId');

    final data = await ApiService.get(params.toString()) as Map<String, dynamic>;
    final rawItems = (data['items'] as List?) ?? [];

    return PagedPatients(
      items: rawItems
          .map((p) => PatientModel.fromJson(p as Map<String, dynamic>))
          .toList(),
      totalCount: data['totalCount'] as int? ?? 0,
    );
  }

  static Future<PatientModel> getById(int id) async {
    final data = await ApiService.get('Patient/$id') as Map<String, dynamic>;
    return PatientModel.fromJson(data);
  }
}

File: pharmion-backend/Pharmion.Services/Services/PatientService.cs
using MapsterMapper;
using Microsoft.EntityFrameworkCore;
using Pharmion.Model.Responses;
using Pharmion.Model.SearchObjects;
using Pharmion.Services.Database;
using Pharmion.Services.Database.Entities;
using Pharmion.Services.Interfaces;

namespace Pharmion.Services.Services
{
    public class PatientService : BaseService<PatientResponse, PatientSearchObject, Patient>, IPatientService
    {
        public PatientService(PharmionDbContext context, IMapper mapper) : base(context, mapper)
        {
        }

        public override async Task<PagedResult<PatientResponse>> GetAsync(PatientSearchObject search)
        {
            var query = _context.Patients
                .Include(p => p.City)
                .AsQueryable();

            if (!string.IsNullOrEmpty(search.Name))
                query = query.Where(p =>
                    (p.FirstName + " " + p.LastName).Contains(search.Name) ||
                    p.Username.Contains(search.Name));

            if (!string.IsNullOrEmpty(search.JMBG))
                query = query.Where(p => p.JMBG.Contains(search.JMBG));

            if (search.IsInsured.HasValue)
                query = query.Where(p => p.IsInsured == search.IsInsured.Value);

            if (search.CityId.HasValue)
                query = query.Where(p => p.CityId == search.CityId.Value);

            int? totalCount = null;
            if (search.IncludeTotalCount)
                totalCount = await query.CountAsync();

            if (!search.RetrieveAll && search.Page.HasValue && search.PageSize.HasValue)
                query = query.Skip(search.Page.Value * search.PageSize.Value)
                             .Take(search.PageSize.Value);

            var patients = await query.OrderBy(p => p.LastName).ToListAsync();

            return new PagedResult<PatientResponse>
            {
                Items = patients.Select(MapToResponse).ToList(),
                TotalCount = totalCount
            };
        }

        public override async Task<PatientResponse?> GetByIdAsync(int id)
        {
            var patient = await _context.Patients
                                     .Include(p => p.City)
                                     .Include(p => p.ChronicDiseases)
                                     .ThenInclude(cd => cd.ChronicDisease)
                                     .FirstOrDefaultAsync(p => p.Id == id);



            return patient == null ? null : MapToResponse(patient);
        }

        private PatientResponse MapToResponse(Patient p) => new PatientResponse
        {
            Id = p.Id,
            FirstName = p.FirstName,
            LastName = p.LastName,
            Username = p.Username,
            Email = p.Email,
            Gender = p.Gender,
            DateOfBirth = p.DateOfBirth,
            JMBG = p.JMBG,
            InsuranceNumber = p.InsuranceNumber,
            Address = p.Address,
            CityId = p.CityId,
            CityName = p.City?.Name ?? string.Empty,
            PhoneNumber = p.PhoneNumber,
            EmergencyContact = p.EmergencyContact,
            IsInsured = p.IsInsured,
            IsActive = p.IsActive,
            CreatedAt = p.CreatedAt,
            ChronicDiseases = p.ChronicDiseases
                                       .Select(cd => cd.ChronicDisease?.Name ?? string.Empty)
                                       .Where(name => !string.IsNullOrEmpty(name))
                                       .ToList(),
        };
    }
}